### PR TITLE
Feature evaluations

### PR DESCRIPTION
**Pull request by SamuelVch98**
Add a page for writing a course evaluation + database storage

### CODE REVIEW

File: course.py [hunk 1]
```diff
@@ -248,3 +248,44 @@ def add_duplicate_course():
         db.session.rollback()
 
     return redirect(url_for('course.course_info', course_id=course_id, current_year=year))
+
+
+@course_bp.route('/evaluations')
+@login_required
+def evaluations():
+    user_id = request.args.get('user_id')
```
Comment (anthonygego) [inline course.py]: It seems you are confused between route and GET arguments. The route processor embeds regex matching and argument existence.

It is therefore more interesting to rely on the route processor for mandatory arguments and on the GET arguments for optional ones.

In this case, user id and current year should probably be part of the route arguments as you would to ensure the argument is specified and is correct.

File: course.py [hunk 2]
```diff
@@ -248,3 +248,44 @@ def add_duplicate_course():
...
+
+@course_bp.route('/create_evaluations', methods=['POST'])
+@login_required
+def create_evaluation():
+    form = request.form
+    if not form:
+        return make_response("Problem with form request", 500)
+
+    course_id = request.form['course_id']
+    current_year = request.args.get('current_year')
```
Comment (anthonygego) [inline course.py]: Same comment here.

File: course.py [hunk 3]
```diff
@@ -248,3 +248,44 @@ def add_duplicate_course():
...
+
+
+@course_bp.route('/create_evaluations', methods=['POST'])
+@login_required
+def create_evaluation():
+    form = request.form
+    if not form:
+        return make_response("Problem with form request", 500)
+
+    course_id = request.form['course_id']
```
Comment (anthonygego) [inline course.py]: Beware, you may get a malformed form and not get the expected keys. Please handle this case.

File: course.py [hunk 4]
```diff
@@ -248,3 +248,44 @@ def add_duplicate_course():
...
+    current_year = request.args.get('current_year')
+    user_id = request.args.get('user_id')
+    tasks = request.form.getlist('tasks[]')
+    evaluation_hour = request.form['evaluation_hour']
+    workload = request.form['workload']
+    comment = request.form['comment']
+    second_course = True if request.form.get('second_course') == 'Yes' else False
+
+    tasks_str = ', '.join(tasks)
```
Comment (anthonygego) [inline course.py]: You may want something more robust and keep a list in a JSON (or equivalent) format. This way the code won't break even if we decide to add a task including a comma.

File: db.py [hunk 5]
```diff
@@ -149,3 +149,23 @@ class CourseOrganization(db.Model):
...
+
+
+class Evaluation(db.Model):
+    __tablename__ = 'evaluation'
+
+    course_id = db.Column(db.Integer, primary_key=True)
+    course_year = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(db.Integer, db.ForeignKey('user.id'), primary_key=True)
+    task = db.Column(db.String)
+    nbr_hours = db.Column(db.String)
```
Comment (anthonygego) [inline db.py]: String ?

**Review by anthonygego — CHANGES_REQUESTED** [review]
Use route arguments for mandatory ones and GET arguments for optional ones only. Please apply this rule in the whole project. We may already have failed to respect it.

File: course.py (hunk 1 above)
Comment (SamuelVch98) [reply course.py]: I understand the difference better now, thank you!

File: course.py [hunk 6]
```diff
@@ -249,3 +249,45 @@ def add_duplicate_course():
...
+
+    try:
+        new_evaluation = Evaluation(course_id=course_id, course_year=current_year, user_id=user_id, task=tasks,
+                                    nbr_hours=evaluation_hour, workload=workload, comment=comment,
+                                    second_course=second_course)
+        db.session.add(new_evaluation)
+        db.session.commit()
+    except Exception as e:
+        db.session.rollback()
+        raise e
```
Comment (anthonygego) [inline course.py]: This exception is propagated but not managed. Actually the form allows to submit an evaluation twice. However the database does not allow it. If one tries to submit the form twice, the exception is thrown.

File: course.py [hunk 7]
```diff
@@ -249,3 +249,45 @@ def add_duplicate_course():
         db.session.rollback()
 
     return redirect(url_for('course.course_info', course_id=course_id, current_year=year))
+
+
+@course_bp.route('/evaluations/<int:user_id>/<int:current_year>')
+@login_required
+def evaluations(user_id, current_year):
+    courses = db.session.query(Course).filter_by(year=current_year).all()
```
Comment (anthonygego) [inline course.py:258]: Maybe the courses that were already evaluated can be filtered out, or, their evaluation can be replaced by a newer one.

File: course.py [hunk 8]
```diff
@@ -249,3 +249,45 @@ def add_duplicate_course():
...
+        new_evaluation = Evaluation(course_id=course_id, course_year=current_year, user_id=user_id, task=tasks,
+                                    nbr_hours=evaluation_hour, workload=workload, comment=comment,
+                                    second_course=second_course)
+        db.session.add(new_evaluation)
+        db.session.commit()
+    except Exception as e:
+        db.session.rollback()
+        raise e
+
+    return redirect(url_for('course.evaluations', user_id=user_id, current_year=current_year))
```
Comment (anthonygego) [inline course.py:303]: You should add some feedback here. Was the form correctly processed ? Was there an issue ? Evaluation already submitted for this course or evaluation replaced by a newer one...

**Review by anthonygego — CHANGES_REQUESTED** [review]

File: course.py [hunk 9]
```diff
@@ -249,3 +250,55 @@ def add_duplicate_course():
...
+
+        new_evaluation = Evaluation(course_id=course_id, course_year=current_year, user_id=user_id, task=tasks,
+                                    nbr_hours=evaluation_hour, workload=workload, comment=comment,
+                                    second_course=second_course)
+        db.session.add(new_evaluation)
+        db.session.commit()
+    except Exception as e:
+        db.session.rollback()
+        flash(f'An error occurred: {str(e)}', 'danger')
+        raise e
```
Comment (anthonygego) [inline course.py]:
```suggestion
```

**Review by anthonygego — APPROVED** [review]

**Pull request merged**